### PR TITLE
Add more half/bf16 conversions.

### LINKED ISSUE #2077: Cast from `__half` to `bool` not supported
### 🐛 Describe the bug

```
import torch
from torch._C._nvfuser import FusionDefinition, Fusion, DataType

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :                           
    T0 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguous=[True, True, True], dtype=DataType.Half)
    T1 = fd.ops.cast(T0, dtype=DataType.Float)                                   
    S2 = fd.define_constant(1)                                               
    T3 = fd.ops.mul(T1, S2)                                                                   
    S4 = fd.define_constant(1.00000) 
    T5 = fd.ops.sub(S4, T3)                                                
    T6 = fd.ops.cast(T5, dtype=DataType.Half)
    T7 = fd.ops.cast(T6, dtype=DataType.Bool)                                                           
    fd.add_output(T7)

inputs = [
    torch.randn(2, 4, 6, device='cuda', dtype=torch.float16),
]

fs = Fusion()
with FusionDefinition(fs) as fd:
    nvfuser_fusion_id0(fd)

for _ in range(5) :
    fs.execute(inputs)
```
Error:

```
RuntimeError: Illegal Cast value from  DataType: __half to DataType: bool
```

### Versions

TOT

### PR DESCRIPTION
**Pull request by csarofeen**
Fixes #2077
@zasdfgbnm please run full tests before merging this one too as I didn't here either.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu [hunk 1]
```diff
@@ -17,18 +17,98 @@ struct __align__(2) __bfloat {
...
       : "=h"(__NVFUSER_BFLOAT_TO_US(val))
       : "f"(f));
   return val;
 }
 
-__device__ float __bfloat2float(const __bfloat h) {
+__device__ __inline__ __bfloat __double2bfloat(const double d) {
+#if __CUDA_ARCH__ >= 700
+  __bfloat val;
+  asm("{  cvt.rn.bf16.f64 %0, %1;}\n"
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu:31]:
```
.relu modifier and {.f16x2, .bf16, .bf16x2, .tf32} destination formats require sm_80 or higher.

cvt.bf16.{u8/s8/u16/s16/u32/s32/u64/s64/f16/f64/bf16}, cvt.{u8/s8/u16/s16/u32/s32/u64/s64/f16/f64}.bf16, and cvt.tf32.f32.{relu}.{rn/rz} require sm_90 or higher.
```
I see we have sm_70 in the requirement. :fearful: This doesn't seem right.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu:31]: Reference: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cvt
`Target ISA Notes`

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu:31]: Thanks, copy and paste got the better of me.

File: torch/csrc/jit/codegen/cuda/test/test_gpu3.cpp [hunk 2]
```diff
@@ -6630,6 +6630,59 @@ TEST_F(NVFuserTest, FusionHuggingFaceRepro2064_CUDA) {
       "");
 }
 
+TEST_F(NVFuserTest, FusionCastings_CUDA) {
+  // TODO: this test is failing
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/test/test_gpu3.cpp:6636]: I added this new test, and it is currently failing

File: torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu [hunk 3]
```diff
@@ -17,18 +17,98 @@ struct __align__(2) __bfloat {
...
   asm("{  mov.b32 %0, {0,%1};}\n"
       : "=f"(val)
       : "h"(__NVFUSER_BFLOAT_TO_CUS(h)));
   return val;
 }
+
+__device__ __inline__ double __bfloat2double(const __bfloat h) {
+#if __CUDA_ARCH__ >= 700
+  double val;
+  asm("{  cvt.f64.f16 %0, %1;}\n"
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu]:
```suggestion
  asm("{  cvt.f64.bf16 %0, %1;}\n"
```

File: torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu [hunk 4]
```diff
@@ -17,18 +17,98 @@ struct __align__(2) __bfloat {
...
+  return val;
+#else
+  return static_cast<double>(__bfloat2float(h));
+#endif
+}
+
+__device__ int __bfloat2int32(const __bfloat h) {
+#if __CUDA_ARCH__ >= 700
+  int val;
+  asm("{  cvt.rzi.s32.f16 %0, %1;}\n"
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu]:
```suggestion
  asm("{  cvt.rzi.s32.bf16 %0, %1;}\n"
```

File: torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu [hunk 5]
```diff
@@ -17,18 +17,98 @@ struct __align__(2) __bfloat {
...
+  return val;
+#else
+  return static_cast<int>(__bfloat2float(h));
+#endif
+}
+
+__device__ __inline__ int64_t __bfloat2int(const __bfloat h) {
+#if __CUDA_ARCH__ >= 700
+  int64_t val;
+  asm("{  cvt.rzi.s64.f16 %0, %1;}\n"
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu]:
```suggestion
  asm("{  cvt.rzi.s64.bf16 %0, %1;}\n"
```

File: torch/csrc/jit/codegen/cuda/runtime/fp16_support.cu [hunk 6]
```diff
@@ -43,7 +37,41 @@ __device__ __half __double2half(const double d) {
...
+  return val;
+#else
+  return __float2half(static_cast<float>(i));
+#endif
+}
+
+__device__ __inline__ __half __int2half(const int64_t i64) {
+#if __CUDA_ARCH__ >= 700
+  __half val;
+  asm("{  cvt.rn.bf16.s64 %0, %1;}\n"
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/runtime/fp16_support.cu]:
```suggestion
  asm("{  cvt.rn.f16.s64 %0, %1;}\n"
```

File: torch/csrc/jit/codegen/cuda/runtime/fp16_support.cu [hunk 7]
```diff
@@ -43,7 +37,41 @@ __device__ __half __double2half(const double d) {
 #endif
 }
 
-__device__ double __half2double(const __half h) {
+__device__ __inline__ __half __int322half(const int i) {
+#if __CUDA_ARCH__ >= 700
+  __half val;
+  asm("{  cvt.rn.bf16.s32 %0, %1;}\n"
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/runtime/fp16_support.cu]:
```suggestion
  asm("{  cvt.rn.f16.s32 %0, %1;}\n"
```

File: torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/runtime/bf16_support.cu:31]: I fixed this

**Review by zasdfgbnm — APPROVED** [review]
Tested on H100 and RTX 4090, both passes

**Pull request merged**